### ISSUE TITLE
WFSS Reference Files for NIRCAM

### ISSUE THREAD

**Issue by sosey** [STScI-JWST/jwst#732]
aXe software uses a text file to relay the dispersion and grism configuration information to the software, and the NIRCAM team has used that similar format,  which  looks like this:
```
# DMS pixels 

BEAM_+1 

# lambda = f(x0,y0,t)
DISPL_+1_0 24000. 
DISPL_+1_1 26000.

DISPX_+1_0 0.
DISPX_+1_1 0.

DISPY_+1_0 -1530.8764939967652
DISPY_+1_1 2589.641434263754

SENSITIVITY_+1 NIRCam.A.1st.sensitivity.fits

XRANGE_+1 -129.74 123.56 
YRANGE_+1 -194.61 657.95

BEAM_+2

DISPL_+2_0 24000. 
DISPL_+2_1 26000.

DISPX_+2_0 0.
DISPX_+2_1 0.

DISPY_+2_0 859.5617529710912
DISPY_+2_1 5179.282868527087

SENSITIVITY_+2 NIRCam.A.2nd.sensitivity.fits

XRANGE_+2 -129.74 123.56
YRANGE_+2 -194.61 657.95
```
 I'm proposing a single asdf file which contains this information, plus what is contained in the sensitivity files, as well as the polynomial models which directly relate to the coefficients being used;  something like this:
```
#ASDF 1.0.0
#ASDF_STANDARD 1.0.0
%YAML 1.1
%TAG ! tag:stsci.edu:asdf/
--- !core/asdf-1.0.0
AUTHOR: Sosey
CHANNEL: SHORT
DESCRIP: C Configuration
DETECTOR: NRCA1
EXP_TYPE: NRC_GRISM
INSTRUMENT: NIRCAM
MODULE: A
PEDIGREE: GROUND
REFTYPE: GRISM
SUBARRAY: ''
TELESCOP: JWST
TITLE: NIRCAM Grism Configuration
USEAFTER: '2014-01-01T00:00:00'
asdf_library: !core/software-1.0.0 {author: Space Telescope Science Institute, homepage: 'http://github.com/spacetelescope/asdf',
  name: asdf, version: 1.2.1}
history:
- !core/history_entry-1.0.0
  description: Initial file creation
  software: !core/software-1.0.0 {author: Sosey, homepage: 'https://github.com/spacetelescope/jwreftools',
    name: nircam_reftools.py, version: '0.7'}
  time: 2017-03-16 23:01:20.483959
spectral_orders:
  '+1':
    DISPL: [24000.0, 26000.0]
    DISPX: [0.0, 0.0]
    DISPY: [-1530.8764939967652, 2589.641434263754]
    MMAG_EXTRACT: 0.0
    MMAG_MARK: 0.0
    SENSITIVITY: NIRCam.A.1st.sensitivity.fits
    SENSITIVITY_ERROR: !core/ndarray-1.0.0
      source: 1
      datatype: float32
      byteorder: big
      shape: [28]
      offset: 8
      strides: [12]
    SENSITIVITY_SENSITIVITY: !core/ndarray-1.0.0
      source: 1
      datatype: float32
      byteorder: big
      shape: [28]
      offset: 4
      strides: [12]
    SENSITIVITY_WAVELENGTH: !core/ndarray-1.0.0
      source: 1
      datatype: float32
      byteorder: big
      shape: [28]
      strides: [12]
    SENSITIVITY_WRANGE: [24500.0, 50000.0]
    XRANGE: [-129.74, 123.56]
    YRANGE: [-194.61, 657.95]
    wx: 0.0
    wy: 0.0
  '+2':
    DISPL: [24000.0, 26000.0]
    DISPX: [0.0, 0.0]
    DISPY: [859.5617529710912, 5179.282868527087]
    MMAG_EXTRACT: 0.0
    MMAG_MARK: 0.0
    SENSITIVITY: NIRCam.A.2nd.sensitivity.fits
    SENSITIVITY_ERROR: !core/ndarray-1.0.0
      source: 0
      datatype: float32
      byteorder: big
      shape: [28]
      offset: 8
      strides: [12]
    SENSITIVITY_SENSITIVITY: !core/ndarray-1.0.0
      source: 0
      datatype: float32
      byteorder: big
      shape: [28]
      offset: 4
      strides: [12]
    SENSITIVITY_WAVELENGTH: !core/ndarray-1.0.0
      source: 0
      datatype: float32
      byteorder: big
      shape: [28]
      strides: [12]
    SENSITIVITY_WRANGE: [24000.0, 50000.0]
    XRANGE: [-129.74, 123.56]
    YRANGE: [-194.61, 657.95]
    wx: 0.0
    wy: 0.0

```
With the above format, a polynomial dispersion model can be attached to each spectral order (if the traces of the orders vary from each other) or attached to all of the orders. That would show up as a 'model' member.

[The calibration page on confluence](https://confluence.stsci.edu/pages/viewpage.action?pageId=61442456) also states that: "We will supply a separate reference file for each combination of filter and grism".  Is there a reason grism+passband need to be linked? We can get the filter information from the direct image which is associated with the grism observations. If the passband data is needed for flat fielding the pixels, we can read in the appropriate reference file and scale the sensitivity array from the grism reference file as needed. This way, if the sensitivity or other grism parameters need to be updated, we only need to redeliver the grism reference file and not all the combinations.

**Comment by sosey** [comment]
Secondary question - do the polynomial dispersion models themselves change with filter+grism combination? If not, we could also save a dictionary of coefficients for each filter in the grism reference file, something like DISPX above would be DISPX[filter_name]=[ list of coefficients ]

**Comment by hbushouse** [comment]
I believe that the reason for the statement "... separate reference file for each combination of filter and grism" does not refer to the direct images in any way, but just the grism exposures, which use a filter and grism in the optical path at the same time (the filter is used to provide low- and high-wavelength cutoffs). So for example you would have grism exposures that use something like F140W+GRISMC, hence the need for reference data that depends on a filter+grism combination.

The reference data for all possible filter+grism combo's could still be put into a single reference file and we'd let the cal pipeline modules find the appropriate entries, but it's probably simplest all the way around to have them in separate files and we'll let the call to CRDS find and return the appropriate one based on filter+grism.

**Comment by hbushouse** [comment]
Is it necessary to list the sensitivity files in this reference file? The sensitivity reference data are used in the photom step and don't really have anything to do with the WCS and spectral trace mapping. I would prefer to have this reference file, used by assign_wcs, contain only WCS-related info. We'll then have separate photom reference files for sensitivity data.

Perhaps the same comment applies to entries like MMAG_EXTRACT and MMAG_MARK. In the aXe world, this one reference file was used to control everything, but in the jwst pipeline we prefer to keep different types of reference data in separate reference files. If MMAG_EXTRACT and MMAG_MARK are used in some other step like extract_2d, then we should figure out a way to move that info (unless there's no convenient way to do so).

**Comment by sosey** [comment]
regarding the filter + grism:  My question was trying to understand why or if any filter information is needed in the grism reference file. I just saw the NIRISS remarks on the calib page though, saying they were creating filter dependent trace solutions, I don't _think_ this is true for nircam.  If the polynomial forms themselves aren't changing with filter, just the coefficients they use, then it may make sense to store the single dispersion model, and keep a dictionary of the changing filter coefficients all in the same file. assign_wcs can then ask for the filter combo that was used and choose the correct set. 

For the MMAG* keywords, is wcs going to mark all detected objects from the catalog, regardless of whether they should be completely ignored? If so, I need to know not to make wcs boxes for them and exclude them in the object list. In that case, since wcs is parsing the catalog and saving information for each object, it may make sense to save a flag with extract or not specified that the later extract_2d step can pick up on. 
```
Sources with magnitudes fainter than the second cutoff (MMAG_MARK_X) are completely ignored.
```
I could see the information storage going either direction. At some level, it's nice to keep everything that has to do with a specific grism together, including maybe the sensitivity information since it's order specific. The MMAG limits are also order specific, and maybe wcs specific. It also may make it easier for users looking at the reference file to have the information in the same place too. But if the reference files should refer not to parameters relating to  an element, but only parameters a step needs, then splitting them makes more sense. Common parameters would either have to be replicated between reference files (eww) or passed along internally.

**Comment by gbrammer** [comment]
1. For the JWST grisms you need to specify a combination of filter and grism because both of these elements are in the optical path for the slitless exposure independent of the direct image that was used (though in practice that will be the same filter).  In the case of NIRISS WFSS, the **grism** is actually specified in the **FILTER** DMS keyword and the blocking filter is in the **PUPIL** keyword.

2. For NIRISS, we do have spectral trace calibrations that are unique for each filter+grism combination because there is a wedge offset between the filters.  The shape of the trace is actually taken to be the same, but the DC offset component of the trace is different.

3. Also remember that there are issues with the NIRISS filter wheel position repeatability such that we also have to keep track of an additional header keyword that is used to rotate the spectral trace polynomials.  

4. I've given my NIRISS files to Nor for him to translate them into the new parameterization format.  Hopefully they'll be available soon.

**Comment by sosey** [comment]
*new* clarification question: The trace solutions take as inputs location in detector pixels in the direct image, and translate from the direct image (distortion corrected) to the grism image (not distortion correct) detector pixel location. However, the grism images are dithered between direct image exposures, and the direct images are drizzle combined before making the object catalog.  

How is the science team assuming we will translate the direct image pixel location for input to the trace polynomials?  The polynomial inputs, as I see them, only accept pixel location from the direct image. 

The only way through this I see is the following:
- start with the ra,dec of objects in the catalog (produced from the drizzled direct images)
- use the wcs information for each grism image to translate those to x,y locations in the grism image
- **assume** that those x,y locations in the grism images map one-to-one with where that object would be in a corresponding direct image, and feed those x,y locations into the trace polynomials

Is this the correct assumption to make?

**Comment by gbrammer** [comment]
Yes, I think that is right.  Is there a reason why the computed x,y grism positions **wouldn't** map one-to-one with the position of a (non-existent) direct image?

**Comment by sosey** [comment]
- actual pointing offset being different than commanded
- funky distortion solution actually happening in the grism image which differs slightly from the distortion calculated in imagine alone due to filter + grism combined bandpass?

More precisely, the catalog is based on a distortion corrected image. Should I take the RA,DEC from the catalog and send it through the imaging mode distortion solution to get the v2/v3 (ra, dec ) location of the object and use that to calculate the detector (x,y) location in the grism? I feel like ignoring the distortion solution will give us the wrong starting location in the grism image.

**Comment by npirzkal** [comment]
Hi

Yes, that sounds right. But there could be reasons why objects are in x,y in the direct image but not in the grism “images” fi there are any kind of filter wedge effects.
I would not assume anything at this stage and we have no idea (I think) of how large this effect would be in the various JWST instruements. With WFC3, we know that the filters introduce a filter wedge offset. This is now accounted for in the grismconf module however. It is handled as a simple x,y shift that is contained as part of the information stored int he configuration file. Since NIRCAM has so many filters, it makes a lot more sense to handle this effect this way rather than generating many grism configuration files as was done in the past for WFC3. It is also why the new polynomials are easier to deal with since a simple x,d offsets can be applied trivially, which was not the case with the old aXe format.
So, to summarize, I would not assume this at this point and would suggest at least building what we have found to be required for the WFC3 grism.

Cheers,
N

**Comment by gbrammer** [comment]
I agree with the points raised in this discussion.   We'll be starting with the assumption that the relative astrometry between the direct image and the dithered grism exposures is reliable, and for now you can treat the distorted WCS of the grism exposures as if it was that of a direct image without the dispersing element.    

One minor point will be to make sure to allow as valid an (x,y) coordinate pair computed from a catalog v2/v3 (ra/dec) position to have both negative values and values larger than the detector axis lengths.

**Comment by npirzkal** [comment]
One small addition is that the exact amount of catalog “padding” to account for the POM being larger than the detectors is something inside the config file at this time. So as long as an area larger than this is allowed when computing the WCS coordinates in the grism ref. frame, the grism code should be able to do the right thing.

N

**Comment by sosey** [comment]
I'm closing this issue, the new reference files needed for the spec2 pipeline were delivered for 7.1 and their current formats can be seen through crds, docs describing them are included with the pipeline docs.
